### ISSUE TITLE
Elasticsearch search fails to give an Result

### ISSUE THREAD

**Issue by crmwolf** [salesagility/SuiteCRM#7742]
**Issue**
The Elasticsearch Engine does not provide any Searchresults within CRM Global Search (Elastic search Engine in vendor/bin/robo gives expected result)

**Expected Behavior**
A Result based on the Input into the Search input field should appear

**Actual Behavior**
Nothing appears on the Search Results Page and this Error Appears

`PHP Catchable fatal error:  Object of class Link2 could not be converted to string in /crm/lib/Search/SearchResults.php on line 213`

**Possible Fix**
Changed:

protected function getLink($label, $module, $record, $action)
    {
        global $sugar_config;
        $link = "<a href=\"{$sugar_config['site_url']}/index.php?action={$action}&module={$module}&record={$record}&offset=1\"><span>{$label}</span></a>";
        return $link;
    }

**to** 

protected function getLink($label, $module, $record, $action)
    {
        global $sugar_config;
        if(!is_string($record)){
                return "";
        }
        $link = "<a href=\"{$sugar_config['site_url']}/index.php?action={$action}&module={$module}&record={$record}&offset=1\"><span>{$label}</span></a>";
        return $link;
    }

Just a Workaround here.

**Steps to Reproduce**
1.  Installing SuiteCRM 7.11.6
2. Installing ElasticSearch
3. Enabling ElasticSearch in CRM and generating an Index
4. Searching for something within the Global Field

**Your Environment**
* SuiteCRM Version used: **7.11.6**
* Browser name and version (e.g. Chrome Version 51.0.2704.63 (64-bit)): **Chrome (Latest)**
* Environment name and version (e.g. MySQL, PHP 7): **PHP 7.1**
* Operating System and version (e.g Ubuntu 16.04): **Ubuntu 18.04.2 LTS**

**Comment by Mac-Rae** [comment]
Confirmed in 7.11.8

**Comment by larsschaps** [comment]
Hi...
I had this bug too...
I think everything is fine, until the method
`$obj->load_relationship()`
is called in `lib\Search\SearchResults::getHitsAsBeans()`.

Before that, certain links with 3 fields in the object $obj in question are fine.
-  _\<Link name\>_ contains the Link2 object
- _<Link_name>\_name_ contains the name of the linked bean.
- _<Link_name>\_ida_ contains the id of the linked bean.

After the call to `SugarBean::load_relationship()`, the field 
`<Link_name>_ida` contains the the Link2 object, as well. I think this is the actual problem later on,

`SugarBean::load_relationship()` calls `SugarBean::get_linked_fields()` returns an array which contains the _\_ida_ fields. Then `SugarBean::load_relationship($rel_name)` is called with the \_ida field name as  the parameter and this will overwrite the ID in the \_ida with a Link2 object, which feels wrong.

My personal workaround was to overwrite the method `get_linked_fields` in my custom beans, so as not to not return _ida fields. But that would not help here. I suggest checking in the foreach loop in `SugarBean::load_relationship()` to see if the current element is an _ida field. 

I could try to send a pull request.

Hope that helps.
Lars

**Comment by tsummerer** [comment]
You're absolutely right about the three fields and the ida field being filled with incorrect data. 

Part of the problem is that if a Relationship is created via studio the id field is (IMO) incorrectly set as a `link` type when it should be an `id` type. 

A relationship _should_ consist of a `link` field to define the relationship, an `id` field to hold the id, and a `relate` field to display the name. 
A studio relationship creates two `link` fields and one `relate`. 
This flaw causes trouble in the API too. 

So the other workaround is to go and update your relevant vardefs and change the _ida fields to `id` type

**Comment by tsitle** [comment]
I believe that the PR #8513 should also fix this issue.

**Comment by johnM2401** [comment]
Hey!

Thank you for raising this issue.

I've given this a try in the latest release, 7.14.6, using elasticSearch7

After indexing:
![Image](https://github.com/user-attachments/assets/7ffcf579-6561-4c98-a94e-6838e3078790)

I am able to search as expected, using the elasticSearch option:
![Image](https://github.com/user-attachments/assets/5ffb0c3e-fc2f-4380-8e53-4d2a2d477806)

<br>

There have been a large number of fixes to elasticsearch functionality since 7.11.8, (Notably in 7.11.19 and 7.12.7).
So it may be that this has since been resolved.

I'm going to close this issue off for now.
However, if you're still seeing this issue in the latest releases of SuiteCRM, please don't hesitate to get in touch!

Thank you for your contributions to the project!
